pip_audit/dependency_source/resolvelib/pypi_provider.py

```diff
@@ -18,19 +21,21 @@
 import requests
 from packaging.requirements import Requirement
 from packaging.specifiers import SpecifierSet
-from packaging.utils import canonicalize_name
-from packaging.version import InvalidVersion, Version
+from packaging.utils import canonicalize_name, parse_sdist_filename, parse_wheel_filename
+from packaging.version import Version
 from resolvelib.providers import AbstractProvider
+from virtualenvapi.manage import VirtualEnvironment  # type: ignore
```

Comment:
We probably want to use the builtin `venv` module provided by Python, since "venv" is the standard solution as of Python 3 and doesn't make as many legacy (i.e. Python 2) accommodations. I'll work that out.

Reply:
I did see that module but couldn't spot a way to inspect packages and things like that. It seemed to be more about creating virtual envs from what I could tell.

Reply:
Yeah, it looks like the public API is pretty barebones. Hmm. We could probably do something like entering the virtual environment and then using `pip-api`, but that's a little circular, so maybe sticking with `virtualenvapi` makes sense for now 🙂

Reply:
Hmm, yeah maybe that is a better idea. I'm starting to wonder if it's a good idea to add `virtualenvapi` as a dependency. 

The part that I'm not sure about is "entering the virtual environment". Seems like I'd have to call into a separate Python via `system` or something like that. Is that your understanding too?

Reply:
> The part that I'm not sure about is "entering the virtual environment". Seems like I'd have to call into a separate Python via system or something like that. Is that your understanding too?

There's a mysterious `context` object in the `venv` API, and I was hoping that something in there would give us access to the temporary environment so that we don't have to do anything roundabout like shelling out to a child Python process. But I haven't had time to look into it yet; could you spend some time investigating that and seeing whether there's some nicety there that we can lean on?

Otherwise: yeah, I think building up a bit of scaffolding for creating a Python subprocess within that virtual environment is the right approach.

Reply:
I don't see anything like that in the `venv` API. There is some code [here](https://github.com/python/cpython/blob/3.9/Lib/venv/__init__.py#L111) that shows how the context gets created, but it doesn't seem to have what we want. From what I can see, the `venv` only really wraps what you can do with `venv` from the command line. So anything like introspecting environments, installing things within them, etc looks to be out of scope for `venv`.

Reply:
I think if we want to remove the `virtualenvapi` dependency, we'll have to write some code that does:
- Creating the virtual environment with `venv`.
- Shell out to Python and do something like:
```
source env/bin/activate
pip install -e sdist_dir/
pip install pip-api
python -c "import pip_api; print(pip_api.installed_packages)"
```
- Parse the shell output into package name and version pairs.
- Filter out `pip_api` dependencies since those will turn up in the output also.

This is all possible but I'm thinking that it might be worth just getting everything working end to end and then having "remove `virtualenvapi` dependency" as a nice-to-have just so we can unblock the other tasks and keep moving.

What do you think @woodruffw.

Reply:
Yeah, let's put this info in an issue and revisit it once we have this initial version in.

Reply:
Captured this in #50.

Reply:
Thanks!